File: gol.pl
#!/usr/bin/env perl

use 5.014;

my @map = ();
my $w = 30;
my $prob = 0.5;

sub init {
    for my $i (0..$w-1) {
        for my $j (0 .. $w-1) {
            push @map, (int rand(100) < $prob * 100);
        }
    }
}

sub get { $map[ $_[0]*$w + $_[1] ] }

sub print_ {
    for my $i (0..$w-1) {
        for my $j (0..$w-1) {
            print substr("  X ", get($i, $j), 2);
        }
        say "";
    }
}

sub neighbors {
    my ($x, $y) = @_;
    my @dirs = ([-1, -1], [0, -1], [1, -1],
                [-1,  0],          [1,  0],
                [-1,  1], [0,  1], [1,  1]);
    my $cnt = 0;
    for my $d (@dirs) {
        $cnt += ($x+$d->[0] >= 0 and $y+$d->[1] >= 0 and
                 $x+$d->[0] < $w and $y+$d->[1] < $w and
                 get($x+$d->[0], $y+$d->[1]));

    }
    return $cnt;
}

sub next_state {
    my @ns = ();
    for my $i (0..$w-1) {
        for my $j (0..$w-1) {
            my $nc = neighbors($i, $j);
            push @ns, (get($i, $j) ? ($nc == 2 or $nc == 3) : ($nc == 3));
        }
    }
    @map = @ns;
}

# main
init();
print_();
while (1) {
    next_state();
    print_();
    say "="x60;
    select undef, undef, undef, 0.1;
}

File: gol.lua
#!/usr/bin/env lua

function init()
    math.randomseed( os.time() )
    -- in Lua, indecs start at 1, not 0
    for i = 1, w do
        map[i] = {}
        for j = 1, w do
            map[i][j] = (math.random() < prob and 1) or 0 
        end
    end
end


function print_()
    for i = 1, w do
        for j = 1, w do
            io.write (((map[i][j] == 1 and 'X') or ' ') .. ' ')
        end
        print ()
    end
end


function neighbors(x, y)
    local dirs = { {-1, -1}, {0, -1}, {1, -1},
                   {-1,  0},          {1,  0},
                   {-1,  1}, {0,  1}, {1,  1} }
    local cnt = 0
    for i = 1, 8 do
        if (x+dirs[i][1] > 0 and y+dirs[i][2] > 0 and
            x+dirs[i][1] <=  w and y+dirs[i][2] <=  w and 
            map[ x+dirs[i][1] ][ y+dirs[i][2]] == 1) then
                cnt = cnt + 1
        end
    end
    return cnt
end


function next_state()
    local ns = {}
    for i = 1, w do
        ns[i] = {}
        for j = 1, w do
            local nc = neighbors(i, j)
            ns[i][j] = (((map[i][j] == 1 and (nc == 2 or nc == 3)) or
                         (nc == 3)) and 1) or 0
        end
    end

    for i = 1, w do
        for j = 1, w do
            map[i][j] = ns[i][j]
        end
    end
end

w = 30
prob = 0.5
map = {}
init()
print_()

while 1 do
    next_state()
    print_()
    local time = os.clock()
    while os.clock()-time < 0.1 do end
end
